core/arch/aarch64/emit_utils.c

```diff
@@ -181,36 +207,22 @@ exit_cti_reaches_target(dcontext_t *dcontext, fragment_t *f, linkstub_t *l,
 void
 patch_stub(fragment_t *f, cache_pc stub_pc, cache_pc target_pc, bool hot_patch)
 {
-    /* Compute offset as unsigned, modulo arithmetic. */
-    ptr_uint_t off = (ptr_uint_t)target_pc - (ptr_uint_t)stub_pc;
-    if (off + 0x8000000 < 0x10000000) {
```

Comment:
@derekbruening For now, I've removed the direct branch patching, as that is not within the architecture specifications. The alternative is to keep it and accept that it isn't strictly okay.

As discussed, I can work on gathering some numbers to help us understand how important this is.

Reply:
You mean, patching a CBZ, CBNZ, TBZ, or TBNZ?  Patching a B or BL is still explicitly on the opcode mod list.  Probably B is the most common in any case?

Reply:
(BL presumably turns into B after mangling)

Reply:
I mean that I've removed the `STP <-> B` mangling that's done by `patch_stub` for near fragment linking. 

Per the manual, both opcodes _before_ and _after_ need to be part of the opcode mod list, of which `STP` is not a part.

We could do `NOP <-> B` by adding an additional `NOP` as the first stub instruction, but the only guarantee we get is that the executed instruction will be _either one of them, not the latest one_. 

To guarantee that only the latest instruction is executed, we'd still need a sync, which also includes all other threads executing an `ISB` after [`clear_icache`](https://github.com/DynamoRIO/dynamorio/blob/57784023fc35b5aad7108c444ab79fc820276543/core/drlibc/drlibc.c#L106); I don't think we have any way to do that right now (#1911).

Reply:
So you mean not the exit cti direct branch patching, but the intermediate patching that went through the stub.  But the exit cti patching could involve a CBZ.  And does this mean that the intermediate is all gone and now there's nothing but exit cti or indirect branch?

Reply:
Maybe file (or add to an issue) sthg about the inlined syscall skipping.  If we have no good solution there, we might decide to live with a lack of theoretical guarantee if no known hardware is like that, and perhaps we could try to get ARM to change the spec.

Reply:
I put some comments in #1911 about the inlined skipping and these other larger concerns.

Reply:
The inlined syscall skip patching is relevant only to the thread that's receiving the signal. So we only need to worry about making the patched code visible to the current processor element.

The current implementation calls machine_cache_sync[1] after patching, which is also specified by this blogpost[2] on ARM website. (Though [2] also mentions some branch predictor ops that we don't do today)

I suppose it is still possible that the thread is moved to some other processor element, in which case it'll execute stale code.

I'll add these details to #1911 for now, as we don't know for sure if this is a concern.

[1]: https://github.com/DynamoRIO/dynamorio/blob/b3bc4dfad113575cdb89730168c1113d090c93b8/core/arch/mangle_shared.c#L816

[2]: https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/caches-and-self-modifying-code